src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionListener.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter.issue;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.engine.reporting.ReportEntry;
+import org.junit.platform.launcher.TestExecutionListener;
+import org.junit.platform.launcher.TestIdentifier;
+import org.junit.platform.launcher.TestPlan;
+
+/**
+ * <p>This listener collects the names and results of all tests, which are annotated with the {@link org.junitpioneer.jupiter.Issue} annotation.
+ * After all tests are finished the results are provided to an {@link IssueProcessor} for further processing.</p>
+ */
+public class IssueExtensionListener implements TestExecutionListener {
+
+	static final String KEY_ISSUE = "Issue";
+
+	// Cache with all tests that belong to an issue <issueId, List<UniqueIdentifier>>
+	private final ConcurrentHashMap<String, List<String>> issueTestsCache = new ConcurrentHashMap<>();
+
+	// Cache with tests results of test cases <UniqueIdentifier, result>
+	private final ConcurrentHashMap<String, String> testStatusCache = new ConcurrentHashMap<>();
+
+	// Package private by purpose for testing
+	List<IssuedTestCase> allIssuedTests = new ArrayList<>();
```

Comment:
Then I would add code just for testing in a productive code base, which is bad coding. Package private variables, methods, etc. are best practices for testable code.

I know that you don't like that (see your comment above about package private method), but I keep this coding style fully minded. I had this discussions for years at my work. I tested my methods and had almost no errors in my code, while colleagues had uncountable amount of errors, because they uses private methods and then claimed, that they could not test their code.